### PR TITLE
Better keyboard

### PR DESCRIPTION
**Pull request by shokifrend007**
Hi, I found the keyboard to be a weak point of this project, so I improved it.

I made the font bigger and made the touch zones have a half-a-letter margin around them (making them bigger).
fbcon.c was modified because it always picked the biggest font available (wich was the 10x18 used by the keyboard).
I also got bottom_lcd.c to accept the bigger font, but any smaller ones will not work anymore.

### Download

[bins.zip](https://github.com/linux-3ds/linux/files/7786460/bins.zip)

### Comparison

**Old (photo not by me)**
image

**New**
image

### CODE REVIEW

**Comment by shokifrend007** [comment]
By the way, if you have anything I can do to make this project better, please let me know.

**Comment by Wolfvak** [comment]
For starters, can you think of any way to bring back the lost keys? There's a lot of notebook keyboard layouts where they still manage to fit them in. Also, doesn't the config need to be modified to build the 10x18 font too?

**Comment by shokifrend007** [comment]
Would this be better? Or should the row be on top?
image

**Comment by Wolfvak** [comment]
I'd say on the top, or maybe the right side as well.

I don't think the function keys are needed for the most part, so maybe try replacing those with Home, etc?

**Comment by Wolfvak** [comment]
Also, I just noticed you're actually modifying the fbcon driver - it'd be better to keep the changes self contained to the tsc driver (since the kernel-level swkbd driver will be replaced eventually anyway...)

**Comment by shokifrend007** [comment]
Hi, i had to change the driver because the kernel always picked the largest font available. But there is a command line option to set the font used by the kernel. Is there any way to compile-in command line options?

Also, why do you want the extra keys so much?

**Comment by Wolfvak** [comment]
> Hi, i had to change the driver because the kernel always picked the largest font available. But there is a command line option to set the font used by the kernel. Is there any way to compile-in command line options?

There are functions to manually blit characters to the bottom screen, you can change them to use a custom font if needed instead of changing `fbcon`.

> Also, why do you want the extra keys so much?

I heavily use PgUp/PgDown, Delete and Home when using it as a "normal linux system" while debugging (`dmesg | less`, going back to the start of the line, etc). I agree they won't be needed when the system is more stable, but for now it's a requirement.

**Comment by shokifrend007** [comment]
I don't know if you understood me correctly but with "the kernel always uses the largest font" I mean, that the top screen (console) used the bigger font, which meant that it was like 20x30 chars or something. That is the reason that I had to rewrite `fbcon` so that it used the smaller font.

**Comment by shokifrend007** [comment]
This is what it looks like after the last commit:
image

**Comment by shokifrend007** [comment]
Please let me know if you know anything about compiling-in command line options, because that would be much cleaner than modifying `fbcon.c`.

**Comment by Wolfvak** [comment]
Aha, you can change the parameters in `the arch/arm/boot/dts/nintendo3ds_*.dts` files.

**Comment by shokifrend007** [comment]
Now, that the  font change is in the `cmdline` instead of in `fbcon.c`, I think this pull request is finished, or are you missing anything?

**Comment by shokifrend007** [comment]
btw: do you have a discord server or something for this project?

**Comment by kleo** [comment]
GodMode9 Discord #3ds-linux https://discord.gg/BRcbvtFxX4

**Comment by shokifrend007** [comment]
ok, thanks

**Comment by Wolfvak** [comment]
Looks good to me!

**Review by Wolfvak — APPROVED** [review]

File: arch/arm/mach-ctr/bottom_lcd.c [hunk 1]
```diff
@@ -92,12 +92,13 @@ void nintendo3ds_bottom_lcd_clear_screen(unsigned int color)
 void nintendo3ds_bottom_lcd_draw_char(const struct font_desc *font, int x, int y, unsigned int color, char c)
 {
 	int i, j;
-	const u8 *src;
+	const u16 *src;
+	const u8 bytes_per_char = 36;//font->charcount / sizeof font->data; // (doesn't work)
```
Comment (xerpi) [inline arch/arm/mach-ctr/bottom_lcd.c:96]: `struct font_desc::data` is a pointer so `sizeof` will return 4.
I think the way to calculate `bytes_per_char` is:
```c
bytes_per_char = ((font->width + 7) / 8) * font->height;
```
Check https://elixir.bootlin.com/linux/latest/source/lib/fonts/font_10x18.c#L5135

But I think this is note needed. Also, I'd prefer if the code was generic for any font size.

**Comment by Wolfvak** [comment]
Thanks for the contribution!

**Comment by Samfitzgamer2022** [comment]
maybe make it easy to install GNOME or something like that

**Pull request merged**